### PR TITLE
[moveonly] Fix CConnman template methods to be fully-defined in net.h

### PR DESCRIPTION
**Pull request by Empact**
Currently template methods `CConnman::ForEachNode` and `CConnman::ForEachNodeThen` rely on `CConnman::NodeFullyConnected`, which is defined in net.cpp. This can result in linker errors like so:
https://travis-ci.org/bitcoin/bitcoin/jobs/379011082

The fix is to move `CNode`'s declaration above` CConnman` in net.h, so that `CConnman` can make use of `CNode` members in the method implementations, and move `NodeFullyConnected` from net.cpp to net.h.

### CODE REVIEW

**Comment by Empact** [comment]
It could be helpful to review 0acdc9e using `git diff --color-moved=zebra --minimal`

**Comment by promag** [comment]
utACK d53c3a6.

**Comment by Empact** [comment]
Rebased

**Comment by sipa** [comment]
utACK ff9e94a65f990905abc48aafae2d5ff0cb61778b

`--color-moved=dimmed_zebra` is very useful, I didn't know about that option.

**Comment by Empact** [comment]
@sipa Agreed! Was just added: https://blog.github.com/2018-04-05-git-217-released/

Hoping github adds support for it at some point. :P

**Comment by Empact** [comment]
@MarcoFalke here's the build failure from #13235. Is there another approach you recommend?

```
libtool: link: /usr/bin/ccache arm-linux-gnueabihf-g++ -std=c++11 -Wstack-protector -fstack-protector-all -fPIE -pipe -O2 -fvisibility=hidden -Wl,--exclude-libs -Wl,ALL -pthread -Wl,-z -Wl,relro -Wl,-z -Wl,now -pie -o bench/bench_bitcoin bench/bench_bench_bitcoin-bench_bitcoin.o bench/bench_bench_bitcoin-bench.o bench/bench_bench_bitcoin-checkblock.o bench/bench_bench_bitcoin-checkqueue.o bench/bench_bench_bitcoin-Examples.o bench/bench_bench_bitcoin-rollingbloom.o bench/bench_bench_bitcoin-crypto_hash.o bench/bench_bench_bitcoin-ccoins_caching.o bench/bench_bench_bitcoin-mempool_eviction.o bench/bench_bench_bitcoin-verify_script.o bench/bench_bench_bitcoin-base58.o bench/bench_bench_bitcoin-lockedpool.o bench/bench_bench_bitcoin-prevector.o bench/bench_bench_bitcoin-coin_selection.o  -L/home/travis/build/bitcoin/bitcoin/depends/arm-linux-gnueabihf/share/../lib libbitcoin_server.a libbitcoin_wallet.a libbitcoin_common.a libbitcoin_util.a libbitcoin_consensus.a crypto/libbitcoin_crypto.a leveldb/libleveldb.a leveldb/libleveldb_sse42.a leveldb/libmemenv.a secp256k1/.libs/libsecp256k1.a univalue/.libs/libunivalue.a libbitcoin_zmq.a -L/home/travis/build/bitcoin/bitcoin/depends/arm-linux-gnueabihf/lib /home/travis/build/bitcoin/bitcoin/depends/arm-linux-gnueabihf/lib/libzmq.a -lpthread -ldl -lboost_system-mt -lboost_filesystem-mt -lboost_program_options-mt -lboost_thread-mt -lboost_chrono-mt -ldb_cxx-4.8 -lssl -lcrypto -ldl -lcrypto -ldl -lminiupnpc /home/travis/build/bitcoin/bitcoin/depends/arm-linux-gnueabihf/lib/libevent_pthreads.a /home/travis/build/bitcoin/bitcoin/depends/arm-linux-gnueabihf/lib/libevent.a -lrt -pthread
libbitcoin_wallet.a(libbitcoin_wallet_a-wallet.o): In function `CWalletTx::RelayWalletTransaction(CConnman*)':
wallet.cpp:(.text+0x5cea): undefined reference to `CConnman::NodeFullyConnected(CNode const*)'
collect2: error: ld returned 1 exit status
```
https://travis-ci.org/bitcoin/bitcoin/jobs/379011082

**Comment by Empact** [comment]
I spent a bit of time going down the path of adding the wallet.cpp sources to libbitcoin_wallet.a, but this ended up seeming like the more narrow fix, because the other approach kept wanting to pull more of the project into the wallet lib.

**Comment by practicalswift** [comment]
Concept ACK

**Comment by Empact** [comment]
Fixed comment location in e2db96e, and squashed the latter two commits.

File: src/net.h [hunk 1]
```diff
@@ -109,754 +148,710 @@ struct CSerializedNetMsg
...
+    int readHeader(const char *pch, unsigned int nBytes);
+    int readData(const char *pch, unsigned int nBytes);
+};
 
-    size_t GetNodeCount(NumConnections num);
-    void GetNodeStats(std::vector<CNodeStats>& vstats);
-    bool DisconnectNode(const std::string& node);
-    bool DisconnectNode(NodeId id);
+/** Information about a peer */
+class CConnman;
```
[the remark this answers is not in the document]
Comment (Empact) [reply src/net.h]: Fixed

**Comment by Empact** [comment]
Unnecessary thanks to build fix now in #13235

**Pull request closed without merging**